contracts/libraries/ProxyDeltas.sol

```diff
@@ -15,68 +15,58 @@ struct ProxyDeltas {
 /// The list works optimally if after applying a series of changes it's iterated over.
 /// The list uses 2 words of storage per stored cycle.
 library ProxyDeltasImpl {
+    using ProxyDeltasImpl for ProxyDeltas;
+
     struct ProxyDeltaStored {
         uint64 next;
         int128 thisCycleDelta;
+        uint64 isAttached;
         // --- SLOT BOUNDARY
         int128 nextCycleDelta;
+        uint128 slotFiller;
     }
 
     uint64 internal constant CYCLE_ROOT = 0;
-    uint64 internal constant CYCLE_END = type(uint64).max;
 
     /// @notice Return the next non-zero, non-obsolete delta and its cycle.
     /// The order is undefined, it may or may not be chronological.
     /// Prunes all the fully zeroed or obsolete items found between the current and the next cycle.
     /// Iterating over the whole list prunes all the zeroed and obsolete items.
-    /// @param current The previously returned cycle or CYCLE_ROOT to start iterating
+    /// @param oldCurrent The previously returned cycle or CYCLE_ROOT to start iterating
+    /// @param oldNext The previously returned `next` or CYCLE_ROOT to start iterating
     /// @param finishedCycle The last finished cycle.
     /// Entries describing cycles before `finishedCycle` are considered obsolete.
-    /// @return next The next iterated cycle or CYCLE_ROOT if the end of the list was reached.
+    /// @return current The next iterated cycle or CYCLE_ROOT if the end of the list was reached.
```

Comment:
When I read through this and the code the generic terminology (old, current, next) made it harder to follow the logic. Especially here where it says “current: the nexter iterated cycle”.

To improve this I’d suggest the following signature
~~~solidity
function nextDeltaPruning(
    ProxyDeltas storage self,
    uint64 prevCycle,
    uint64 cycleHint,
    uint64 finishedCycle
)
    internal
    returns (
        uint64 cycle,
        uint64 cycleHint,
        int128 thisCycleDelta,
        int128 nextCycleDelta
    )
~~~

Reply:
Great idea with the `Hint` convention, applied